lib/fizzy/instantiate.cpp

```diff
@@ -305,7 +305,8 @@ ExternalGlobal find_imported_global(const std::string& module, const std::string
     return {it->value, module_global_type};
 }
 
-std::optional<uint32_t> find_export(const Module& module, ExternalKind kind, std::string_view name)
+std::optional<uint32_t> find_export(
```

Comment:
Are we sure that none of the constructors used would throw through `std::make_optional`? Can clang-tidy verify that at all?

Reply:
Perhaps we could add a test marked noexcept constructing all those, just to have an early indication if we brake something that clang-tidy can't detect.

Reply:
I've added `noexcept` to some constructors more, and I'm pretty sure, all `External*` objects use either explicitly `noexcept` or default move-constructors, which are noexcept implicitly.

Reply:
> Perhaps we could add a test marked noexcept constructing all those, just to have an early indication if we brake something that clang-tidy can't detect.

How would we detect it in test, if clang-tidy doesn't detect it? Well setting `terminate_handler` might work...

Reply:
If you're worried about `make_optional` only, it's used only with `uint32_t`.
Other functions construct an object and then directly construct `optional` return value.

Reply:
I'm just not convinced anymore that anything wrong is detected when we mark stuff noexcept 😕

Reply:
Yeah now it seems to me too, that clang-tidy doesn't check much.

Reply:
Also apparently `std::find_if` can throw: https://en.cppreference.com/w/cpp/algorithm/find

Reply:
I guess it makes sense to use manual loop instead of `find_if` then.

Reply:
> Also apparently `std::find_if` can throw: https://en.cppreference.com/w/cpp/algorithm/find

Actually, it's only overloads with `ExecutionPolicy` can throw.

Reply:
I think regular `find_if` is not `noexcept`, because the predicate function is allowed to throw.

I think it's safe to leave `noexcept` with `find_if` here, but I made labmda `noexcept`, too.

Reply:
> > Also apparently std::find_if can throw: https://en.cppreference.com/w/cpp/algorithm/find
>
> Actually, it's only overloads with ExecutionPolicy can throw.

I am not sure, that description is kind of confusing. Makes me think it can throw `bad_alloc` in any case.

Reply:
First of all `noexcept` is not like `const` that all internal function calls must be `noexcept`. Usage of `find_if` looks fine to me.

> Makes me think it can throw `bad_alloc` in any case.

Only overloads with `ExecutionPolicy` which we don't use.

`clang-tidy` only reports obvious mistakes (as direct usage of `throw` inside `noexcept`). We can dig deeper if you know the check name that fails.